2022/Q2/artifacts/pos-model/PoS-model.md

```diff
@@ -184,9 +201,85 @@ change_consensus_key(validator_address, consensus_key)
   validators[validator_address].consensus_key[cur_epoch+pipeline_length] = consensus_key
 }
 ```
-## Delegator transactions:
+## Delegator transactions
+
+It is essentially a copy and paste of the validator transactions with minor changes. Once we converge on the validator transactions, we can propagate the changes to the delegator transactions.
+
+```go
+delegate(validator_address, delegator_address, amount, offset_length)
+{
+  //add amount bond to delta at n+pipeline_length
```

Comment:
> I was checking their [bond_tokens transaction](https://github.com/anoma/anoma/blob/c366704610f1f823fc696815e843f5a4ab3431ea/proof_of_stake/src/lib.rs#L280), and I cannot find where they check this. It maybe that it is somehow implicit when they get the validators state. I say we ask this in the slack channel directly.

https://github.com/anoma/anoma/blob/c366704610f1f823fc696815e843f5a4ab3431ea/proof_of_stake/src/lib.rs#L1457 I think here is the check on the validator state.  (this is called from https://github.com/anoma/anoma/blob/c366704610f1f823fc696815e843f5a4ab3431ea/proof_of_stake/src/lib.rs#L314 )

Reply:
> https://github.com/anoma/anoma/blob/c366704610f1f823fc696815e843f5a4ab3431ea/proof_of_stake/src/lib.rs#L1457 I think here is the check on the validator state. (this is called from https://github.com/anoma/anoma/blob/c366704610f1f823fc696815e843f5a4ab3431ea/proof_of_stake/src/lib.rs#L314 )

Thanks @jmalicevic, I missed that :) I'll add it to the model.

Reply:
> OK, three comments. (a) It's strange they have the check in the transaction code and not in the VP, since the VP's correctness shouldn't depend on the tx. This is probably worth adding a comment in the model to raise their attention.

I guess they do not have safety checks completely decoupled from the transactions. Maybe there are things like this that have to be checked by the transaction: probably the transaction cannot be executed at all if the validator does not exist. I think in the case of PoS transactions, this should be still OK: PoS state should not be accessible by user-defined transactions. In any case, from the model point of view, where things are checked does not matter much, as we are putting it all together in the model.

(b) We need to adapt our model to reflect that `become_validator` can only be called once. 

Done.

(c) What if we mess up and lose our rewards key? Or want to change it for some reason? Isn't this a plausible use case?

They do not seem to allow that. I guess one can create a new validator in any of those cases.